#131: Introduce 'modify' to adjust historical records to fixed date/time.

sruffell opened the pull request.

lauft commented on test/modify.t:
```diff
@@ -0,0 +1,158 @@
+        code, out, err = self.t("modify stop @2 {:%Y-%m-%dT%H:%M:%S}Z".format(now_utc - timedelta(hours=2)))
+
+        # Verify that the total time is one hour less than it was before.
+        code, out, err = self.t("summary")
```
Better use `self.t.export()` than parsing the output of `timew summary`, see e.g. `move.t:117`:
```
j = self.t.export()

self.assertEquals(len(j), 2)
self.assertClosedInterval(j[0],
                          expectedStart="20170301T110000Z",
                          expectedEnd="20170301T133000Z",
                          expectedTags=["foo"])
```

sruffell replied:
Ack

lauft commented on test/modify.t:
```diff
@@ -0,0 +1,158 @@
+
+        code, out, err = self.t("modify start @2 {:%Y-%m-%dT%H:%M:%S}Z".format(now_utc - timedelta(hours=2)))
+
+        code, out, err = self.t("summary")
```
See comment at line 86

sruffell replied:
Ack

lauft commented on test/modify.t:
```diff
@@ -0,0 +1,158 @@
+
+        code, out, err = self.t.runError("modify start @1 {:%Y-%m-%dT%H:%M:%S}Z".format(now_utc - timedelta(hours=2)))
+        self.assertIn("You cannot overlap intervals", err)
+
```
Test `modify should not move start beyond end`?
Test `modify should not move end before start`?

sruffell replied:
I believe these case are already covered?  I've updated the test description to be of this form in the updated pull request.

lauft commented on src/commands/CmdModify.cpp:
```diff
@@ -0,0 +1,95 @@
+  if (filter.start.toEpoch() == 0)
+    throw std::string ("No updated time specified. See 'timew help modify'.");
+
+  database.startTransaction ();
```
Maybe move this line down to line 85? Start transaction right before the database stuff happens...

sruffell replied:
ack

sruffell commented on src/commands/CmdHelp.cpp:
```diff
@@ -52,6 +52,7 @@ int CmdHelpUsage (const Extensions& extensions)
             << "       timew help [<command> | interval | hints | date | duration]\n"
             << "       timew join @<id> @<id>\n"
             << "       timew lengthen @<id> [@<id> ...] <duration>\n"
+            << "       timew modify (start|stop) @<id> <date>\n"
```
Ack. I was thinking from a user standpoint I want to modify the argument passed to the `stop` command and that end was internal usage, but I see it referenced in the user documentation.

sruffell commented on src/commands/CmdHelp.cpp:
```diff
@@ -693,6 +694,26 @@ int CmdHelp (
+                << '\n'
+                << "  $ timew modify stop @2 2018-10-10T17:10:00\n"
+                << '\n'
+                << "See also 'summary', 'lengthen', and 'shorten'.\n"
```
Heh, I was not aware of the resize command. I see it in the man pages (which I also see that I did not update for the modify command) but it is not in the output of help.

But in general, ack. I will make this change in the updated pull request.

sruffell commented:
Thanks for review. I've updated the branch with what I hope are the fixes to all your comments.

Additionally I've:

- Updated the tests to make sure only the start of an open interval can be modified by the end cannot.
- Updated the man page for the `modify` command.

With regards to the man pages, I was not sure of the proper formatting for the non-optional argument to the modify command. I also did not see how the individual command pages get run through the configure with the cmake files. So there is a good chance I missed something there.

sruffell commented on src/commands/CmdHelp.cpp:
```diff
@@ -693,6 +694,26 @@ int CmdHelp (
+    //              12345678901234567890123456789012345678901234567890123456789012345678901234567890
+    else if (words[0] == "modify")
+      std::cout << '\n'
+                << "Syntax: timew modify (start|stop) @<id> @<date>\n"
```
Should be fixed now.

sruffell commented on src/commands/CmdModify.cpp:
```diff
@@ -0,0 +1,95 @@
+
+  Interval interval = tracked.at (tracked.size () - id);
+  if (interval.is_open ())
+    throw format ("Cannot modify open interval @{1}.", id);
```
Fixed.

lauft commented:
Closes #131
